File: Project Euler/3.py
def is_prime(n: int) -> bool:
    return 0 not in [n % i for i in range(2, n)]


def get_biggest_prime_factor(n: int) -> int:
    if is_prime(n):
        return None
    current: int = 1
    while not n % 2:
                n /= 2
                current = 2
    for i in range(3, round(n**0.5) + 1, 2):
        if is_prime(i):
            while not n % i:
                n /= i
                current = i
    if n > 2:
        current = n
    return current


def main():
    num = 600851475143
    print(get_biggest_prime_factor(num))


if __name__ == "__main__":
    main()

File: Project_Euler/3.cpp
#include <iostream>
using namespace std;

bool isPrime(int n) {
    if (n <= 1) {
        return false;
    }
    if (n == 2) {
        return true;
    }
    if (n % 2 == 0) {
        return false;
    }
    for (int i = 3; i * i <= n; i += 2) {
        if (n % i == 0) {
            return false;
        }
    }
    return true;
}

int getBiggestPrimeFactor(long n) {
    if (isPrime(n)) {
        return n;
    }
    int biggestPrimeFactor = 1;
    while (n % 2 == 0) {
        n /= 2;
        biggestPrimeFactor = 2;
    }
    for (int i = 3; i * i <= n; i += 2) {
        while (n % i == 0) {
            n /= i;
            biggestPrimeFactor = i;
        }
    }
    if (n > 2) {
        biggestPrimeFactor = n;
    }
    return biggestPrimeFactor;
}

// what the actual fuck
// how can python be so slow

int main(int argc, char const *argv[]) {
    long num = 600851475143;
    cout << getBiggestPrimeFactor(num) << endl;
    return 0;
}
